caffe2/core/tensor.cc

```diff
@@ -201,9 +201,16 @@ void Tensor::enforce_invariants() {
   if (impl_.get() == nullptr) {
     throw std::runtime_error("TensorImpl with nullptr is not supported");
   }
-  CAFFE_ENFORCE(
-      !impl_->is_variable(),
-      "Caffe2 tensor wrapper doesn't support autograd variables");
+  // TODO: only check `!impl_->requires_grad()` after Variable and Tensor are merged
+#ifndef CAFFE2_FB_LIMITED_MOBILE_CAPABILITY
+    CAFFE_ENFORCE(
+      !impl_->is_variable() || (impl_->is_variable() && !(impl_->requires_grad() && torch::autograd::GradMode::is_enabled())),
+      "Caffe2 tensor wrapper doesn't support autograd variables that require grad");
+#else // defined(CAFFE2_FB_LIMITED_MOBILE_CAPABILITY)
+    CAFFE_ENFORCE(
+      !impl_->is_variable() || (impl_->is_variable() && !(impl_->requires_grad())),
```

Comment:
we don't have GradMode on mobile or something?  That seems...annoying to keep trac kof.

Reply:
@ljk53 told me that `thread_local` should work on mobile. After https://github.com/pytorch/pytorch/pull/18573 I will try to remove the spacial casing and see if we run into any build error.